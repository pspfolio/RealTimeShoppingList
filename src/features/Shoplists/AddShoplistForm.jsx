import React, { Component } from 'react';
import styled from 'styled-components';

const FlexWrapper = styled.div`
  display: flex;
  flex-direction: column;
  flex: 1;
`;

class ShoplistForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      name: '',
      category: 'fruitsvegetables'
    };

    this.handleAddItem = this.handleAddItem.bind(this);
    this.handleCategoryChange = this.handleCategoryChange.bind(this);
  }

  handleAddItem(event) {
    const { name, category } = this.state;
    const { onItemAdd } = this.props;

    onItemAdd({ name, category });

    this.setState({ name: '', category: '' });
    event.preventDefault();
  }

  handleCategoryChange(event) {
    const { target } = event;

    this.setState({ category: { value: target.value, name: target.options[target.selectedIndex].text } });
  }

  render() {
    const { name, category } = this.state;
    const { title, onTitleChange } = this.props;

    return (
      <FlexWrapper>
        <label htmlFor="name">
          Shoppinglist name
          <input id="name" type="text" value={title} onChange={onTitleChange} />
        </label>

        <label htmlFor="item">
          Shoplist item
          <input id="item" type="text" value={name} onChange={event => this.setState({ name: event.target.value })} />
        </label>

        <label htmlFor="category">
          Pick category
          <select value={category.value} onChange={this.handleCategoryChange} id="category">
            <option value="fruitsvegetables">Fruits & Vegetables</option>
            <option value="bread">Bread</option>
            <option value="milkyogurts">Milk & Yogurts</option>
            <option value="meat">Meat</option>
            <option value="drinks">Drinks</option>
          </select>
        </label>

        <button type="button" onClick={this.handleAddItem}>
          Add item
        </button>
      </FlexWrapper>
    );
  }
}

export default ShoplistForm;
