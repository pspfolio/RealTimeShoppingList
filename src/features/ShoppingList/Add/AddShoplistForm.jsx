import React, { Component } from 'react';
import styled from 'styled-components';
import map from 'lodash/map';
import TextField from 'material-ui/TextField';
import RaisedButton from 'material-ui/RaisedButton';
import { purple500 } from 'material-ui/styles/colors';
import { database } from '../../../firebase/index';
import ShoplistCategorySelectField from '../ShoplistCategorySelectField';

const FlexWrapper = styled.div`
  display: flex;
  flex-direction: column;
  flex: 1;
`;

const styles = {
  underlineStyle: {
    borderColor: purple500
  },
  floatingLabelFocusStyle: {
    color: purple500
  }
};

class ShoplistForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      name: '',
      category: { value: 'fruitsvegetables', name: 'Fruits & Vegetables' },
      categories: []
    };

    this.handleAddItem = this.handleAddItem.bind(this);
    this.handleCategoryChange = this.handleCategoryChange.bind(this);
    this.refCategories = database.ref('categories');
  }

  componentDidMount() {
    this.refCategories.once('value', snap => {
      const categories = map(snap.val(), (category, key) => ({ value: key, name: category }));
      this.setState({ categories });
    });
  }

  handleAddItem(event) {
    const { name, category } = this.state;
    const { onItemAdd } = this.props;

    onItemAdd({ name, category });

    this.setState({ name: '', category: { value: 'fruitsvegetables', name: 'Fruits & Vegetables' } });
    event.preventDefault();
  }

  handleCategoryChange(event, index, value) {
    const { categories } = this.state;
    const { name } = categories.filter(category => category.value === value)[0];
    this.setState({ category: { value, name } });
  }

  render() {
    const { name, category, categories } = this.state;
    const { title, onTitleChange } = this.props;
    return (
      <FlexWrapper>
        <TextField
          hintText="Title of shoplist"
          floatingLabelText="Title"
          value={title}
          onChange={onTitleChange}
          underlineFocusStyle={styles.underlineStyle}
          floatingLabelFocusStyle={styles.floatingLabelFocusStyle}
        />

        <TextField
          hintText="Item name"
          floatingLabelText="Item"
          value={name}
          onChange={event => this.setState({ name: event.target.value })}
          underlineFocusStyle={styles.underlineStyle}
          floatingLabelFocusStyle={styles.floatingLabelFocusStyle}
        />

        <ShoplistCategorySelectField
          categories={categories}
          handleCategoryChange={this.handleCategoryChange}
          value={category.value}
        />

        <RaisedButton
          type="button"
          label="Add item"
          onClick={this.handleAddItem}
          backgroundColor={purple500}
          labelColor="white"
          disabled={name === ''}
        />
      </FlexWrapper>
    );
  }
}

export default ShoplistForm;
