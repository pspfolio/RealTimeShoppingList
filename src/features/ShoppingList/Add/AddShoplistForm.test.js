import React from 'react';
import { shallow } from 'enzyme';
import RaisedButton from 'material-ui/RaisedButton';
import AddShoplistForm from './AddShoplistForm';

test('renders component', () => {
  const wrapper = shallow(<AddShoplistForm />);
  expect(wrapper).toMatchSnapshot();
});

test('add button is disabled by default', () => {
  const wrapper = shallow(<AddShoplistForm />);
  expect(wrapper.find(RaisedButton).prop('disabled')).toBeTruthy();
});

test('Should enable add button on name state change', () => {
  const wrapper = shallow(<AddShoplistForm />);
  expect(wrapper.find(RaisedButton).prop('disabled')).toBeTruthy();
  wrapper.setState({ name: 'bar' });
  expect(wrapper.find(RaisedButton).prop('disabled')).toBeFalsy();
});
