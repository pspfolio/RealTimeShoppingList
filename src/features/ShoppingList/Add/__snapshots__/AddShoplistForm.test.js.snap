// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders component 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ShoplistForm />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <TextField
          disabled={false}
          floatingLabelFixed={false}
          floatingLabelFocusStyle={
            Object {
              "color": "#9c27b0",
            }
          }
          floatingLabelText="Title"
          fullWidth={false}
          hintText="Title of shoplist"
          multiLine={false}
          onChange={undefined}
          rows={1}
          type="text"
          underlineFocusStyle={
            Object {
              "borderColor": "#9c27b0",
            }
          }
          underlineShow={true}
          value={undefined}
        />,
        <TextField
          disabled={false}
          floatingLabelFixed={false}
          floatingLabelFocusStyle={
            Object {
              "color": "#9c27b0",
            }
          }
          floatingLabelText="Item"
          fullWidth={false}
          hintText="Item name"
          multiLine={false}
          onChange={[Function]}
          rows={1}
          type="text"
          underlineFocusStyle={
            Object {
              "borderColor": "#9c27b0",
            }
          }
          underlineShow={true}
          value=""
        />,
        <CategorySelectField
          categories={Array []}
          handleCategoryChange={[Function]}
          value="fruitsvegetables"
        />,
        <RaisedButton
          backgroundColor="#9c27b0"
          disabled={true}
          fullWidth={false}
          label="Add item"
          labelColor="white"
          labelPosition="after"
          onClick={[Function]}
          primary={false}
          secondary={false}
          type="button"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "disabled": false,
          "floatingLabelFixed": false,
          "floatingLabelFocusStyle": Object {
            "color": "#9c27b0",
          },
          "floatingLabelText": "Title",
          "fullWidth": false,
          "hintText": "Title of shoplist",
          "multiLine": false,
          "onChange": undefined,
          "rows": 1,
          "type": "text",
          "underlineFocusStyle": Object {
            "borderColor": "#9c27b0",
          },
          "underlineShow": true,
          "value": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "disabled": false,
          "floatingLabelFixed": false,
          "floatingLabelFocusStyle": Object {
            "color": "#9c27b0",
          },
          "floatingLabelText": "Item",
          "fullWidth": false,
          "hintText": "Item name",
          "multiLine": false,
          "onChange": [Function],
          "rows": 1,
          "type": "text",
          "underlineFocusStyle": Object {
            "borderColor": "#9c27b0",
          },
          "underlineShow": true,
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "categories": Array [],
          "handleCategoryChange": [Function],
          "value": "fruitsvegetables",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "backgroundColor": "#9c27b0",
          "disabled": true,
          "fullWidth": false,
          "label": "Add item",
          "labelColor": "white",
          "labelPosition": "after",
          "onClick": [Function],
          "primary": false,
          "secondary": false,
          "type": "button",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <TextField
            disabled={false}
            floatingLabelFixed={false}
            floatingLabelFocusStyle={
              Object {
                "color": "#9c27b0",
              }
            }
            floatingLabelText="Title"
            fullWidth={false}
            hintText="Title of shoplist"
            multiLine={false}
            onChange={undefined}
            rows={1}
            type="text"
            underlineFocusStyle={
              Object {
                "borderColor": "#9c27b0",
              }
            }
            underlineShow={true}
            value={undefined}
          />,
          <TextField
            disabled={false}
            floatingLabelFixed={false}
            floatingLabelFocusStyle={
              Object {
                "color": "#9c27b0",
              }
            }
            floatingLabelText="Item"
            fullWidth={false}
            hintText="Item name"
            multiLine={false}
            onChange={[Function]}
            rows={1}
            type="text"
            underlineFocusStyle={
              Object {
                "borderColor": "#9c27b0",
              }
            }
            underlineShow={true}
            value=""
          />,
          <CategorySelectField
            categories={Array []}
            handleCategoryChange={[Function]}
            value="fruitsvegetables"
          />,
          <RaisedButton
            backgroundColor="#9c27b0"
            disabled={true}
            fullWidth={false}
            label="Add item"
            labelColor="white"
            labelPosition="after"
            onClick={[Function]}
            primary={false}
            secondary={false}
            type="button"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "disabled": false,
            "floatingLabelFixed": false,
            "floatingLabelFocusStyle": Object {
              "color": "#9c27b0",
            },
            "floatingLabelText": "Title",
            "fullWidth": false,
            "hintText": "Title of shoplist",
            "multiLine": false,
            "onChange": undefined,
            "rows": 1,
            "type": "text",
            "underlineFocusStyle": Object {
              "borderColor": "#9c27b0",
            },
            "underlineShow": true,
            "value": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "disabled": false,
            "floatingLabelFixed": false,
            "floatingLabelFocusStyle": Object {
              "color": "#9c27b0",
            },
            "floatingLabelText": "Item",
            "fullWidth": false,
            "hintText": "Item name",
            "multiLine": false,
            "onChange": [Function],
            "rows": 1,
            "type": "text",
            "underlineFocusStyle": Object {
              "borderColor": "#9c27b0",
            },
            "underlineShow": true,
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "categories": Array [],
            "handleCategoryChange": [Function],
            "value": "fruitsvegetables",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "backgroundColor": "#9c27b0",
            "disabled": true,
            "fullWidth": false,
            "label": "Add item",
            "labelColor": "white",
            "labelPosition": "after",
            "onClick": [Function],
            "primary": false,
            "secondary": false,
            "type": "button",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
